drivers/resolver/ad2s1210/ad2s1210.c

```diff
@@ -0,0 +1,361 @@
+	if (ret)
+		return ret;
+
+	dev->mode = mode;
+	return 0;
+}
+
+static int32_t ad2s1210_set_resolution(struct ad2s1210_dev *dev,
+				       uint8_t resolution)
+{
+	int32_t ret;
+	uint8_t control;
+
+	ret = ad2s1210_reg_read(dev, AD2S1210_REG_CONTROL, &control);
+	if (ret)
+		return ret;
+
+	control &= ~(AD2S1210_REG_CONTROL_RES_MASK);
+	switch (resolution) {
+	case 10:
+		control |= 0x0;
+		break;
+	case 12:
+		control |= 0x1;
+		break;
+	case 14:
+		control |= 0x2;
+		break;
+	case 16:
+		control |= 0x3;
+		break;
+	default:
+		pr_err("Invalid resolution: %d\n", resolution);
+		return -EINVAL;
+	}
+
+	ret = ad2s1210_reg_write(dev, AD2S1210_REG_CONTROL, control);
+	if (ret)
+		return ret;
+
+
+	if (!dev->have_resolution_pins)
+		return 0;
+
+	ret = no_os_gpio_set_value(dev->gpio_res0, control & NO_OS_BIT(0));
+	if (ret)
+		return ret;
+
+	ret = no_os_gpio_set_value(dev->gpio_res1, !!(control & NO_OS_BIT(1)));
+	if (ret)
+		return ret;
+
+	return 0;
+}
+
+static int32_t ad2s1210_gpio_init(struct no_os_gpio_desc **desc,
+				  struct no_os_gpio_init_param *param,
+				  uint8_t value,
+				  bool is_optional)
+{
+	int32_t ret;
+
+	if (is_optional && !param->platform_ops)
+		return 0;
+
+	if (is_optional)
+		ret = no_os_gpio_get_optional(desc, param);
+	else
+		ret = no_os_gpio_get(desc, param);
+	if (ret)
+		return ret;
+
+	if (!desc)
+		return 0;
+
+	ret = no_os_gpio_direction_output(*desc, value);
+	if (ret)
+		return ret;
+
+	return 0;
+}
+
+static void ad2s1210_gpio_remove(struct ad2s1210_dev *dev)
+{
+	if (dev->gpio_a0)
+		no_os_gpio_remove(dev->gpio_a0);
+
+	if (dev->gpio_a1)
+		no_os_gpio_remove(dev->gpio_a1);
+
+	if (dev->gpio_res0)
+		no_os_gpio_remove(dev->gpio_res0);
+
+	if (dev->gpio_res1)
+		no_os_gpio_remove(dev->gpio_res1);
+
+	if (dev->gpio_sample)
+		no_os_gpio_remove(dev->gpio_sample);
+}
+
+/*******************************************************************************
+ * @brief Initialize the ad2s1210 driver and create a descriptor.
+ *
+ * @param dev - Device descriptor to create.
+ * @param init_param - Initialization parameters.
+ *
+ * @return Returns negative error code or 0 in case of success.
+ *         Example: -EINVAL - Bad input parameters.
+ *                  -ENOMEM - Failed to allocate memory.
+ *                  0 - No errors encountered.
+*******************************************************************************/
+int32_t ad2s1210_init(struct ad2s1210_dev **dev,
+		      struct ad2s1210_init_param *init_param)
+{
+	struct ad2s1210_dev *d;
+	int32_t ret;
+
+
+	d = (struct ad2s1210_dev *)no_os_calloc(1, sizeof(*d));
+	if (!d)
+		return -ENOMEM;
+
+	d->name = ad2s1210_device_name[0];
+	d->mode = MODE_CONFIG;
+	d->resolution = init_param->resolution;
+
+	ret = ad2s1210_gpio_init(&d->gpio_a0, &init_param->gpio_a0,
+				 NO_OS_GPIO_HIGH, true);
+	if (ret)
+		goto err;
+
+	ret = ad2s1210_gpio_init(&d->gpio_a1, &init_param->gpio_a1,
+				 NO_OS_GPIO_HIGH, true);
+	if (ret)
+		goto err;
```

Comment:
each gpio remove should be handle separately. here your remove all gpios, including the ones that were not initialized yet.

Reply:
i think perhaps the gpio_remove is a bit confusing. in fact the idea was that the ones that are not initialized will  be "null" and will not be removed on the gpio_remove function. ill do individual removes.